sdk/core/core/inc/az_const_str.h

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#ifndef AZ_CONST_STR_H
+#define AZ_CONST_STR_H
+
+#include <az_assert.h>
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// A constant string.
+typedef struct {
+  // Points to the first character.
+  char const *begin;
+  // A numer of C characters (bytes) in the string.
+  size_t size;
+} az_const_str;
+
+typedef struct {
+  char* begin;
+  size_t size;
+} az_str;
+
+// A size of the string literal.
+// Details: to make sure that `S` is a `string literal`, we are appending `""` to `S`.
+#define AZ_STRING_LITERAL_SIZE(S) (sizeof(S "") - 1)
+
+#define AZ_CONST_STR(STRING_LITERAL) \
+  (az_const_str){ .begin = STRING_LITERAL, .size = AZ_STRING_LITERAL_SIZE(STRING_LITERAL) }
+
+inline char const *az_const_str_ptr(az_const_str const buffer, size_t const index) {
```

Comment:
this needs documentation to tell if it's guaranteed to return a null terminated string, and how you may use such a string.

If it's not then this function is going to be bug prone, since a user will have to remember to do `buffer.size - index` to figure out how long the returned buffer is.